core-processor/src/main/java/io/micronaut/inject/beans/visitor/MapperVisitor.java

```diff
@@ -53,63 +54,98 @@ public void visitClass(ClassElement element, VisitorContext context) {
-                                        if (beanProperties.isEmpty()) {
-                                            throw new ProcessingException(element, "@Mapping(from=\"" + from + "\") specifies a property that doesn't exist in type " + fromType.getName());
-                                        }
-                                    }
-                                });
+                    boolean anyMatch = false;
+                    for (ParameterElement parameter: parameters) {
+                        if (parameter.getName().equals(argumentName)) {
+                            anyMatch = true;
+                            if (parameter.getType().getName().equals(Map.class.getName())) {
+                                break;
                             }
+                            if (!hasPropertyWithName(parameter.getGenericType(), propertyName)) {
+                                throw new ProcessingException(element, "@Mapping(from=\"" + from + "\") specifies property " + propertyName + " that doesn't exist in type " + parameter.getGenericType().getName());
+                            }
+                            break;
                         }
-                        if (lastClassElement !=  null) {
-                            lastClassElement.annotate(Mapper.class);
-                        }
+                    }
+                    if (!anyMatch) {
+                        throw new ProcessingException(element, "@Mapping(from=\"" + from + "\") specifies argument " + argumentName + " that doesn't exist for method");
                     }
                 } else {
-                    throw new ProcessingException(element, "Only a single parameter is permitted on a mapping method");
+                    String propertyName = from.substring(from.indexOf(".") + 1);
+                    for (ParameterElement parameter: parameters) {
+                        if (parameter.getType().getName().equals(Map.class.getName())) {
+                            continue;
+                        }
+                        if (!hasPropertyWithName(parameter.getGenericType(), propertyName)) {
+                            throw new ProcessingException(element, "@Mapping(from=\"" + from + "\") specifies property " + propertyName + " that doesn't exist in type " + parameter.getGenericType().getName());
+                        }
+                    }
                 }
-            } else {
-                throw new ProcessingException(element, "@Mapper can only be declared on abstract methods");
-            }
+            });
         }
     }
 
+    private boolean hasPropertyWithName(ClassElement element, String propertyName) {
+        return element.getBeanProperties(PropertyElementQuery.of(element).includes(Collections.singleton(propertyName)))
+            .stream().anyMatch(v -> !v.isExcluded());
```

Comment:
@graemerocher Is this the expected behavior for `PropertyElementQuery` to actually include properties, but set `isExcluded = false`? I was quite confused by this.

Reply:
good question, @dstepanov would know

Reply:
Don't remember, maybe there was a use case where you need to see all the properties